{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 402,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy  as np\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn           as sns \n",
    "\n",
    "import yfinance          as yf\n",
    "import talib\n",
    "\n",
    "from sklearn.neighbors       import KNeighborsClassifier\n",
    "from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score, TimeSeriesSplit\n",
    "from sklearn.preprocessing   import StandardScaler, MinMaxScaler\n",
    "from sklearn.linear_model    import LogisticRegression\n",
    "from sklearn.metrics         import confusion_matrix, classification_report, roc_curve, roc_auc_score\n",
    "from sklearn.svm             import SVC\n",
    "from sklearn.ensemble        import RandomForestClassifier"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Loading Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 414,
   "metadata": {},
   "outputs": [],
   "source": [
    "ibov_df =  pd.read_csv('ibov60.csv', index_col= 'Datetime', parse_dates=True)\n",
    "ibov_df.drop(['retorno', 'media', 'no_trend', 'lag1', 'Close'], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 415,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Open</th>\n",
       "      <th>High</th>\n",
       "      <th>Low</th>\n",
       "      <th>Adj Close</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Datetime</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2019-03-26 10:00:00-03:00</th>\n",
       "      <td>93679.71875</td>\n",
       "      <td>94848.31250</td>\n",
       "      <td>93679.71875</td>\n",
       "      <td>94764.74219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2019-03-26 11:00:00-03:00</th>\n",
       "      <td>94725.67188</td>\n",
       "      <td>94763.88281</td>\n",
       "      <td>93987.67188</td>\n",
       "      <td>94566.67969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2019-03-26 12:00:00-03:00</th>\n",
       "      <td>94577.82031</td>\n",
       "      <td>94929.57031</td>\n",
       "      <td>94344.11719</td>\n",
       "      <td>94753.86719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2019-03-26 13:00:00-03:00</th>\n",
       "      <td>94784.11719</td>\n",
       "      <td>95066.54688</td>\n",
       "      <td>94714.65625</td>\n",
       "      <td>94909.07031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2019-03-26 14:00:00-03:00</th>\n",
       "      <td>94985.82813</td>\n",
       "      <td>95303.38281</td>\n",
       "      <td>94924.22656</td>\n",
       "      <td>95244.81250</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  Open         High          Low    Adj Close\n",
       "Datetime                                                                     \n",
       "2019-03-26 10:00:00-03:00  93679.71875  94848.31250  93679.71875  94764.74219\n",
       "2019-03-26 11:00:00-03:00  94725.67188  94763.88281  93987.67188  94566.67969\n",
       "2019-03-26 12:00:00-03:00  94577.82031  94929.57031  94344.11719  94753.86719\n",
       "2019-03-26 13:00:00-03:00  94784.11719  95066.54688  94714.65625  94909.07031\n",
       "2019-03-26 14:00:00-03:00  94985.82813  95303.38281  94924.22656  95244.81250"
      ]
     },
     "execution_count": 415,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ibov_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Feature Engineering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 416,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create Feature Open - Low\n",
    "ibov_df['Min_Low'] = ibov_df['Open'] - ibov_df['Low']\n",
    "\n",
    "# Bulding a Trend Intensity Indicator \n",
    "period = 60\n",
    "sma = talib.SMA(ibov_df['Adj Close'].values,timeperiod=period)      #Turn down the profit\n",
    "ibov_df['statement'] = ibov_df['Adj Close'] > sma\n",
    "ibov_df['tii'] = (ibov_df['statement'].rolling(period).sum())/period*100\n",
    "ibov_df.drop(['statement'], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 417,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA34AAAFQCAYAAADgEKx7AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nOzdd3RVVd7G8e+9NzeVFNIhoQWQGoqg0kEUscKIYhdlBuziOKMOUoaiFBsWsAy+jhUdiorgjICAIAIjIk2a9ACBFNJ7bjnvH5E7RtoFk5yU57MWi5ySc57sTYBfzj57WwzDMBAREREREZFay2p2ABEREREREalcKvxERERERERqORV+IiIiIiIitZwKPxERERERkVpOhZ+IiIiIiEgtp8JPRERERESklvMxO4CIiNQ+M2fOZNasWafsDwoKonHjxlx33XXce++92O32Ss/icrn45JNPGDJkCIGBgec8v6CggPfee4+vv/6apKQkDMMgJiaGXr16MXz4cOLj4ys9sxnuvvtuNmzYwA8//EBISIjZcUREpIKp8BMRkUpzxRVX0KZNG6CsAMvPz2fjxo28+OKLbN269bTFYUX761//yldffcWgQYPOee6JEye44447SEpKolu3btx00034+Piwd+9e5syZw/z583nrrbfo0aNHpeeuajfeeCOXXnopfn5+ZkcREZFKoMJPREQqzZVXXsmQIUPK7TMMgwcffJCvv/6a9evX071790rNkJGR4fW5L774IklJScyaNYsBAwaUO7Z+/XpGjBjBU089xcqVK/H19a3oqKb6bT+JiEjtonf8RESkSlksFk+R8cMPP5icprxvvvmGuLi4U4o+gO7du3P11VeTnp7Opk2bTEgnIiJy4VT4iYhIlbPZbACnfWq2fv16hg8fTpcuXejUqRO33norS5YsOeW8pKQkHnvsMS6//HLat29P//79mThxIunp6Z5zWrVqxYYNGwC45JJLuPvuu8+ay+l0kpmZyYkTJ057/IEHHuCNN97goosuKrf/6NGjjB07lj59+tCxY0euu+46/vnPf+JwOMqdl56ezsSJE+nTp48n8wsvvEB+fn6580aPHk2rVq3IyclhwoQJ9OzZk8TERIYMGcLSpUtPyZWcnMyECRO48sorSUxMpHPnzgwZMoRPPvmk3HkzZ86kVatWrF+/nqFDh9K+fXsGDhxIQUEBd999N61atSI3N9dzvtvt5uOPP+YPf/gDHTp0oEuXLgwfPpy1a9eetR1FRKT6sU2cOHGi2SFERKR22bBhAxs2bODKK6/0vON3kmEYzJgxg6SkJMaOHUt4eLjn2Pz583nssccoLCxk4MCBdO7cmR07djB37lzsdjtdu3YFIDMzkzvuuIPdu3fTt29fevXqhWEYfPnll6xatYpbb73VU1wmJyeTl5fHyJEj6dGjxyl5fm3Hjh3s3r2bZcuWYbFYiIqKIjg42HM8IiKChIQEAgICPPv27NnDbbfdxubNm7nkkkvo06cPqampfP755xw7dszz9PDYsWPceuutfP/991x66aX0798fp9PJwoUL+e677xg0aJBnspvly5eze/du1q1bx549e7jqqqtISEhg/fr1/Pvf/6Zbt240bNgQKCs6b775ZrZv307Pnj3p3bs3MTEx/PDDDyxfvpz69evToUOHcv3y3XffERUVxYABA4iOjmbgwIF8/vnnJCcnc9999+Hn54fb7ebxxx/nnXfeISAggKuvvppGjRrx3//+lwULFpS7roiIVH96x09ERCrN8uXLSU5OBsoKvoKCAjZs2MDevXsZP348LVq08JybkpLC5MmTSUhIYM6cOdSvXx+Axx9/nHvvvZdXX32V/v37c9FFF/Gf//yHY8eOMXXqVG666SbPNSZPnsycOXNYu3Yt/fr149FHH2XDhg2eguZcs1WOHTuW3bt3k5SUxJQpU5gyZQqNGjXi0ksvpV+/fvTr1++Up5STJk0iJyeH1157jauuusrztY4YMYKFCxcybNgw2rVrx8SJE0lNTeXNN9/k8ssv93z+Bx98wJQpU5g1axZPPfVUuWvbbDa+/PJLz2yk3bt354knnmDevHmeInj27NlkZWXx7rvvlpt05q677mLo0KF8+eWX3HXXXeWu26BBA95//32s1jMP/Fm0aBFLliyhV69ezJw505PhyJEj3H777UydOpU+ffrQqFGjs7apiIhUDxrqKSIilWbFihXMmjWLWbNm8frrr/Pee++xc+dOAgMDyc3NxeVyec5dtGgRpaWljBo1ylP0Afj7+zNq1Cjcbjeff/45UDYEEWDr1q3lrvH444/z3Xff0a9fvwvKGxMTw8KFC3n88cdJSEgAygqdTz/9lEcffZSrr76adevWec5PSUlh48aN9OjRw1P0Qdl7jH/5y1945JFH8PX1JS0tjW+//Za+ffuWK/qgrEBr0KABn3322Sl57rzzznJLUPTt2xeAQ4cOefYNGjSIKVOmnDLTaIcOHfD39z/t5DYDBgw4a9EHeNp64sSJ5TI0atSIBx980PO0UkREagY98RMRkUozbdq0crNFFhYWcuDAAV577TVmzJjBoUOHmDZtGgDbt28Hyt7x27t3b7nrFBYWArB7924ABg4cyOuvv87cuXNZtmwZvXr1ok+fPvTt25eoqKjflTkwMJAHHniABx54gKNHj7J+/XrWrVvHqlWrSE5O5oEHHuBf//oXbdu25eeffwagU6dOp1ynXbt2tGvXDoBVq1ZhGAbZ2dnMnDnzlHPtdjvHjx8nNTWVmJgYz/5mzZqVO+/ksNPS0lLPvq5du9K1a1eys7PZtWsXhw8f5uDBg2zZsoWSkpJyhfFJcXFx52yH3bt3ExMTc9onel26dPGcIyIiNYMKPxERqTKBgYG0b9+eWbNmceWVV/LZZ58xcuRIEhISyMvLA+Bf//rXGT8/JycHKHsyt2DBAt58801WrFjB4sWLWbx4MXa7nSFDhjB27NgKWY8uPj6eoUOHMnToULKzsxk9ejTffPMN77//Ps8995wnT7169c56nZMTpmzZsoUtW7ac8bzs7Oxyhd9vh5VaLBagbCjpSTk5OUybNo0vv/wSh8OBxWIhLi6Obt26sXPnztPex9/f/6x5AfLz84mMjDztsejoaACKi4vPeR0REakeal3hl5+fz2233cZbb71FfHz8Gc87cOAAEyZMICcnh6ioKGbMmEFoaGgVJhURqbt8fX3p3LkzS5Ys4eeffyYhIcEznHD58uVevTfWqFEjpk6disvlYvv27axZs4bPPvuMuXPnEhwczJNPPnlemRYuXMiMGTP461//yuDBg085HhYWxuTJk+nduzdJSUkAnswFBQWnnO92uyktLcXf399z3kMPPcRjjz12XrnO5cknn2T16tXcdtttDB48mIsuushTiC5evPiCrxsUFERaWtppj50seMPCwi74+iIiUrVq1Tt+W7du5fbbby/37sPpnFw8eOTIkSxatIg2bdowe/bsqgkpIiLA/56CnRy+2KpVKwB++umnU849dOgQzz33HCtXrgTK3h2cOHEi+fn52Gw2OnbsyCOPPMKcOXMA+PHHH887T/369UlNTWXZsmVnPOfkE7eTT7xOZt62bdsp527evJlOnTrx5ptves47OZz1t1577TVmz55dbginN3Jzc1m9ejXt27dn0qRJXHzxxZ6i7+jRo5SUlJR7Ong+WrduTW5uLnv27Dnl2MaNGwHKTc4jIiLVW60q/ObNm8eECRM8/yBD2U9wb7zxRgYPHsyYMWMoKSlhx44dBAYG0qdPH6BsXaY777zTrNgiInXO1q1b2bBhA6GhoZ7ZKQcNGoTNZuOVV14ptxaf0+nkmWee4Z///CfZ2dlA2aiNTz755JR16k7OIHpyqQPAs0TCb9fU+61evXrRtGlTli9fzj/+8Q/PBDInlZaWMn36dABuvPFGoOypY+fOnfnuu+9Ys2aN51y3283bb7+NYRj07NmTRo0acckll/Dtt9+esibhwoULef3111mzZs1p1zU8G7vdjtVqJTc3t1zRWFxczDPPPOPV130mJ9/NnDJliucdSyib7Ob111/Hbrdz3XXXXdC1RUSk6tWqoZ5Tpkwpt713717mzZvHv/71L/z8/HjppZd45513aNq0KZGRkYwZM4Zdu3aRkJDA+PHjTUotIlJ7/Xo5BwCXy8W+fftYtWoVLpeLMWPGeN43a9q0KU8++STTp0/n+uuvp3///oSGhvLtt9+yf/9+Lr/8cgYNGgTALbfcwrx583jxxRfZsGEDrVq1IiMjgyVLlhAYGMh9993nuefJd+bGjBlDz549GTZs2Gmz2mw23njjDe655x5mzJjB3Llz6dmzJ+Hh4WRkZLBmzRpSUlIYPnx4uZk5J02axF133cX999/PlVdeSVxcHP/973/ZuXMnw4YN86x1N3nyZO68804ee+wx+vTpQ8uWLTl48CCrVq0iLCyMCRMmnHf7BgQEMGDAAJYuXcrQoUPp2bMnhYWFfPPNN5w4cYLQ0FDy8vJwu93nnMXztwYPHszKlStZunQpgwYNok+fPhQWFrJixQry8vIYP348jRs3Pu/MIiJijlpV+P3W999/T1JSErfccgtQ9lPPtm3bEh8fz4YNG/joo49ITEzklVdeYfr06Z6f5IqISMVYsWIFK1as8Gzb7XbCw8O5/PLLufvuu7n00kvLnT98+HASEhL45z//ybJly3C73TRq1IjRo0dz55134uNT9s9WaGgoH330EW+++SZr167lv//9L/Xq1aNPnz488sgjtGzZ0nPNBx54gP3797N27VoOHTp0xsIPoHnz5ixZsoQPP/yQVatW8fXXX5Ofn09oaCgdOnTgmWee8YwWOalVq1bMnz+fmTNnsnbtWvLz84mPj+fpp58ud6+EhAQ+++wz3njjDVavXs369euJjo5m8ODBPPzwwxe8Ht7UqVOJjY1l+fLlfPTRR0RFRZGYmMh9993Hl19+yfvvv8/3339P9+7dz+u6FouFV155hTlz5rBgwQIWLFhAQEAAnTp14k9/+hPdunW7oLwiImIOi3Ghg/+rsf79+/PBBx+wYsUKjhw5wrhx44Cyl+9dLhc7duxg2rRpLFq0CIB9+/YxatQo/vOf/5gZW0REREREpFLUqnf8fuuyyy7j66+/JiMjA8MwmDhxIu+//z6dO3cmMzPTs/7QypUrPWstiYiIiIiI1Da1eqhn69ateeSRR7jnnntwu920adOG++67Dz8/P15//XXGjRtHUVERsbGxPP/882bHFRERERERqRS1cqiniIiIiIiI/E+tHuopIiIiIiIiKvxERERERERqPRV+IiIiIiIitVytmtwlK6sAt9u8VxYjIuqRkZFv2v3rOrV/9aB+MJ/6wFxqf3Op/c2nPqge1A/mMavtrVYL9esHnfF4rSr83G7D1MLvZAYxj9q/elA/mE99YC61v7nU/uZTH1QP6gfzVMe211BPERERERGRWk6Fn4iIiIiISC2nwk9ERERERKSWU+EnIiIiIiJSy6nwExERERERqeVU+ImIiIiIiNRyVVr45efnc/3113P06NFTji1fvpzBgwczaNAgHnroIXJycqoymoiIiIiIyDkZRvVbqsEbVVb4bd26ldtvv51Dhw6dciw/P5+JEycye/ZsFi1aRKtWrZg5c2ZVRRMRERERETmn9Owinvt4Mx8u+9nsKOetygq/efPmMWHCBKKjo0855nA4mDBhAjExMQC0atWK48ePV1U0ERERERGRM/r5cBYT/7mBv721nj1HsokI8Tc70nnzqaobTZky5YzH6tevz4ABAwAoLi5m9uzZ3H333VUVTUREREREBIDDqXn8sDuNTXvSycwrwel043IbWC0WOjaP4NruTWgZH2Z2zPNWZYWfN/Ly8nj44Ydp3bo1N95443l/fkREvUpIdX6iooLNjlCnqf2rB/WD+dQH5lL7m0vtbz71QfWgfjg/hmGwcVcqk9/9AYCmDUIYcEkMAf4++NptXNG1MVH1A7y6VnVs+2pT+KWlpfGnP/2Jbt26MWbMmAu6RkZGPm63eS9bRkUFk56eZ9r96zq1f/WgfjCf+sBcan9zqf3Npz6oHtQPZ+Z0uVm95RhrfzpOalYhLreBy2Xg+qWO8LFZ+cstHWndpP5vPtHpVZua1fZWq+WsD8KqReHncrl44IEHuOaaa3jooYfMjiMiIiIiIrXU8o1HmffNPgAuahRG09hg7D5WbFYLfr42ureLJayen8kpK56phd/IkSMZNWoUKSkp7Ny5E5fLxdKlSwFo3779Wd8LFBEREREROR87DmZ6ir4XH+pBeA2cpOVCVXnht3LlSs/Hb7/9NgCJiYns3r27qqOIiIiIiEgt5XYbrN+Rwt6j2VitVnYczCA9uxiAK7rE16miD6rJUE8REREREZGK4nC6mfjuBo5nFAIQEmjH5TZo3yycP13fltAgX5MTVj0VfiIiIiIiUqs89/EmjmcUYvex8vyDPepkofdbVbaAu4iIiIiISFU4nlFIkL8P0+7rpqLvFyr8RERERESk1kjPLqKoxFkn3+M7GxV+IiIiIiJSKxiGwaRfFmAPrYVLMvweesdPRERERERqrMJiJzkFJRQWO1m7PYXCEifd2sXQu0MDs6NVKyr8RERERESkxnA4XazacowNu1JxON0cTs0vdzw8xI9BPZvhY9Pgxl9T4SciIiIiIjVCUkoeL83dQn6RA4CmscFcFB9Kp5ZRNIgIJDzEn0bR9UxOWT2p8BMREREREVPtT87hcGoeJQ43pQ4XJQ4XpQ532e9OFyWlLkqdbtKyisgvcvCH3s0YeGlj/Ow2s6PXGCr8RERERESkyjldbt79z24yc4v5+Uh2uWM+Ngt+dhu+v/zys1vxtduIjQjkii7xXH1ZY5NS11wq/EREREREpNIYhkFSah5b9p4gv8hBSamL3EIHPx3IAMDuY6Vfp4ZcdWljQoN88bVbsVn1fl5FU+EnIiIiIiIV7ufDWSz74Qj7knPIKyx7Jy/I3wd/Xxt+vj60bhxGu2bhXHVJY+w+KvQqmwo/ERERERGpUD/+nM7rn/8EQJOYYOKj6nHXVRfRICLI5GR1lwo/ERERERH53Q6n5nHweC5Ol8HHX+8B4JbLW+h9vGpChZ+IiIiIiFywTXvSWbjmIEfTy6+nd3nnOBV91YgKPxERERER8UpuQSnFpU7PtgG8vXgnJQ4XV3aJp0urKBpEBuFjtRLgp6UWqhMVfiIiIiIick47DmXy0r+2nPbYvde0pk/HhlWcSM6HCj8REREREQEgp6CU7QcyKHW6cTjd5BSUcOh4HkfS8skvKpuZ85pujYmL/N8kLT42K51aRJoVWbykwk9EREREpA7ZeSiTklIXFquF/ck5pGUVeYq6XUlZp5wf5O9DYkIEDSKDSGgYQrum4VUdWSqACj8RERERkTpi3jf7WPL94VP2t4gLBQu0iA+lSUww13Vvgq+PFbuPFR+bFYvFYkJaqUgq/EREREREajnDMJj3zT6WbjgCwKNDEgkL9gMgLjIIX7smYqntVPiJiIiIiNRiJaUunv9kEweP59E0NpgnbutEoL/d7FhSxVT4iYiIiIjUcIZhcDyjEJfbwDAMckpcZGYW4DYMDhzL5eDxPBpF1+PJ2zsT4KcSoC5Sr4uIiIiI1ACFxU427EolO78El9vA5TZwuw1SMgvZtj/jrJ9bP9iPv91xsYq+Okw9LyIiIiJSze1KyuKFTzZ7tm1WC1arBdsvv9o0qU/zuBAaRwdjsVioHxZAbl4xVkvZcgst4kL1Hl8dp8JPRERERKSayMwt5p//2cWJnGLyCktxu8HlNnC63ABc260JQ/omYD3HLJtRUcGkp+dVRWSpIVT4iYiIiIhUA0kpeXywdDcHj+fRuWUkESER2GxlBZ4FCz0SY4mLDNLSCnJBVPiJiIiIiFSx1KxCvt+ZyoFjuRhG2XZaVhFQ9j7eg39oj4/NanJKqU1U+ImIiIiIVJLiUie5hQ4cDhcOl5vcglL+vT6JvUdzAPCz22gYGUiQvw8XXxTFrf1bEBHij9Wqp3pSsaq08MvPz+e2227jrbfeIj4+vtyxXbt2MXbsWAoKCujatSuTJk3Cx0d1qYiIiIjUPHmFpXy8fC/f70w97fHIUH+GX9OaixqHYbPqyZ5UviqrrLZu3cq4ceM4dOjQaY8/+eSTPPvss3Tq1IkxY8Ywb9487rjjjqqKJyIiIiLyu2XkFPPTwQw+/novTpeb4EA7vTs0pHFMPew+Vnx9bESG+hMZ5q+CT6pUlRV+8+bNY8KECTz11FOnHEtOTqa4uJhOnToBMGTIEF577TUVfiIiIiJS7bncbhxON8dOFPLsBxsBCPDz4ZEhiXRoHmFyOpEyVVb4TZky5YzH0tLSiIqK8mxHRUWRmnr6x+JnExFR74KyVaSoqGCzI9Rpav/qQf1gPvWBudT+5lL7m68u9EFWXjEf/mcXKzceweU2yh0bfn07buidgN3H3Cd6daEfqqvq2PbV4iU6t9tdblpawzAuaJrajIx83L/5xqtKWi/FXGr/6kH9YD71gbnU/uZS+5uvtvdBSmYh81buY8u+E559f+jdDF8fG3YfK6FBvnRpFUV2VoGJKWt/P1RnZrW91Wo564OwalH4xcbGkp6e7tk+ceIE0dHRJiYSEREREYH121PYlZTl2f7up+MAdG0dTY/2sVwUH0agf7X4L7XIWVWLP6VxcXH4+fnx448/0qVLF7744gv69OljdiwRERERqSNKSl0UO1y43QYpmYWs2pzMgWM5ZOSWABAe4geUrbHXvV0sN/drbmZckfNmauE3cuRIRo0aRWJiIi+++CLjxo0jPz+fdu3aMWzYMDOjiYiIiEgdkZpZyNi3v8dtlH9lqEPzCLq0iqZvp4Y0iAgyKZ1Ixajywm/lypWej99++23Px61bt2bBggVVHUdERERE6qDDqXms3JRMXmEpm/eWva83oGsjGkYG4mOz0r5ZOKH1/ExOKVJxqsVQTxERERGRynQip4iNu9NxGwYHjuWyaU/Z/BL1g/2ICQ/k1v4t6NQi0uSUIpVHhZ+IiIiI1HpLvz/Cik1HPdv1AuyMu6cr0WEBJqYSqTpnLPyefvppry8ybdq0CgkjIiIiIlKRcgpK+WFXKis2HSWmfgCT/ngpFosFm82C9QKWDxOpqc5Y+KWkpHg+drlcbNiwgZiYGNq2bYvdbmfXrl0cO3aMAQMGVElQEREREZHzNefrPWzcnYaf3UaP9rH42m1mRxIxxRkLv3fffdfz8fTp04mLi2Py5MnY7XagbJH1yZMnU1RUVPkpRURERETO09qfjrNxdxpxUUFMuPcSfGxWsyOJmMard/zmz5/P/PnzPUUfgMViYdiwYQwZMoTp06dXWkAREREREW/835c7OZqW79k+nJZPg4hAHhzcXkWf1HleFX4BAQHs37+fhISEcvu3bdtGWFhYpQQTEREREfm1/CIHDqcbp8uNy21QWOzE7TYocbpYsGo/SSl5AJ7ZOcND/OnXuSENI7UGn4hXhd8tt9zC2LFj2b9/P23btsUwDDZv3swHH3zAqFGjKjujiIiIiNRRPx/O4st1h0jPLiYt++yvGLVPCOfuq1oRpZk6RU7hVeH36KOPYrPZ+OijjzhxomyBy5iYGB577DHuueeeSg0oIiIiInXHtv0ZbNqTBsCOg5lk5JYAZevtXXNZY6LrB2CzWvGxWfCxWfH3s2HBQsPIIOoHa8F1kTPxqvCzWCw8/PDDPPzww2RmZmKxWKhfv35lZxMRERGRWs7pcvPV94c5mpbP0fR8jmcUAhBWzxeAixqFMfL6tkSE+psZU6TG83oB95ycHObOncvBgwd54oknWLJkCS1btqR58+aVmU9EREREaomsvBI+WLKb1KwiikqcFJU4KXW6PcfjooLo3i6G/l3iad4w1MSkIrWPV4XfwYMHufPOOwkODiY5OZmHHnqIpUuX8vTTT/POO+9w8cUXV3ZOEREREamBHE432/af4FhGIYu+O4jLbdA+IZzwYD8C/HwI8POhXoCd7u1iCfDz+pmEiJwnr767pk2bxsCBA5kwYQKdO3cG4KWXXmLs2LG89NJLzJkzp1JDioiIiEjNkVdYyp4jOeQWlLB2ewoHjuUC4GOz8tTtnWjdRK8MiVQ1rwq/rVu38re//a3cPqvVyn333ceQIUMqJZiIiIiI1BylDhdpWUV8vGIfy3847NlvAZrHhfDHa9sQFGAnJNDXvJAidZjXz9NLSkpO2ZeRkYGvr755RUREROoSwzBY8v1hkk8UePat356C8cvHfnYbt13Rgo4tIgkOtGOzavF0EbN5Vfj179+fV155hZdfftmz78iRI0ydOpV+/fpVVjYRERERqUZSswqZs2wP2w9mevZF/jLbZkSoP/FR9bi6ZzOig30Jq6elFUSqE68Kv6effpqRI0dy2WWX4XQ6GTp0KDk5OXTs2JGnnnqqsjOKiIiISCUzDAOny6DUWTZk8/udqThcZTNuFhU72bo/g6ISp+f8m/omMPDSxvjYyj/Ni4oKJj09r0qzi8i5eVX4hYSEMHfuXNatW8euXbuw2+20bNmS7t27V3Y+EREREalgbsOg1OFi16Esth3IAGDz3hPkFpSWO69egN3zsWEYDOjaiE4tIriocZiGb4rUMF4VfsOGDWPWrFn06NGDHj16ePZnZmYyYsQIPvvss0oLKCIiIiIXxjAMfj6cTbHDhctl4HK7WftTCvuSsykqcXnOCw3yxeF0Exxo57ruTQnwtdGqcRjR9QNNTC8iFemMhd+mTZs4fLhsRqYffviBRYsWUa9evXLn7Nu3j0OHDlVqQBERERE5P0UlTj5bfYC9R7M5nJZ/yvEmscEkJkQQ4GsjsXkE8VH1TnMVEalNzlj4Wa1Wxo0bh2GUzc80bdq0csctFgtBQUE89NBDlZtQRERERM7K+cu7eMczCnlz4XZSMgsBCPDzITzEjxHXtSXQ3web1fLLPn8z44qICc5Y+HXq1Int27cDZbN6LliwgPDw8CoLJiIiIiJnVlTi5Mv1h1iz9Tj5RY5yx/p1jqN14zAuaR2NxWIxJ6CIVCteveO3ckcxwt0AACAASURBVOXKMx5LSUkhNja2wgKJiIiIyP8czyjgnX/vIiuv/JrKJ7ftPlZu7N3MU+BFhPrTvZ3+byYi5XlV+B05coTnnnuOPXv24HKVvQhsGAalpaVkZmayc+fOSg0pIiIiUle4DYMv1hwkNauQgiIHOw5lARAdFkCrxmHlzm0QEcTlF8fhZ7eZEVVEahCvCr+JEyeSnJzMDTfcwD/+8Q9GjhxJUlISX331FZMnT67sjCIiIiK1lsvt5ufD2bjcBlaLhez8EhavOwRA4+h69OnYkF4dGpDQMASrhm2KyAXyqvDbvHkzs2fPpmvXrnzzzTf07duXTp06kZCQwIoVKxg6dGhl5xQRERGpNYpKnPx8JJvdSVks++HIac8Ze3cXmseFVnEyEamtvCr8nE4ncXFxADRr1ozdu3fTqVMnbrjhBj755JNKDSgiIiJSkzmcLg4ezyOv8OTi6BZ+2J3Khl1pnnOuuDiey9rFYBgGbreBr91G09hgcwKLSK3kVeHXpEkTtm7dSoMGDWjWrJlnts+ioiIKCwsrNaCIiIhITbXnSDYvz9tKicN1yrE2Tepzy+UtCK3nS1g9PxPSiUhd4lXhd8cddzB69GjcbjcDBw7kxhtvJCAggB9//JGOHTt6fbPFixfz5ptv4nQ6ueeee7jzzjvLHd+xYwd///vfcTgcNGjQgBdeeIGQkJDz+4pEREREqgGH08Ur88uKvh7tY7nqkkaeY4YB0fUDCPDz6r9iIiK/m1d/29x+++2Eh4cTHh5Oy5YtmTJlCh9++CGRkZGMHz/eqxulpqby8ssv89lnn+Hr68ttt93GZZddRosWLTznTJkyhVGjRtG3b1+mT5/OO++8w+OPP35hX5mIiIhIJStxuPhy3SFO5BSTnl1EVFgApQ4XuYWlHDyWh9swuLFPAjf0aGp2VBGp47z+MdPAgQM9Hw8ePJjBgwef143WrVtHt27dCAsL81xvyZIlPPLII55z3G43BQUFQNkw0tBQvdAsIiIi1cfupCw+WbGXwmInxaVOCoqdAAT6+VDicJFbUIq/r43gQF8uaxtNk9gQeiVqTT0RMd8ZCz9vn+QBPPPMM+c8Jy0tjaioKM92dHQ027ZtK3fO6NGj+eMf/8jUqVMJCAhg3rx5XmcQERERqSgut5vcAgcpmYUcO1HA1v0nsFosbNufgY/NyqVtovH3teHv60NkqD99OzX0LKAuIlIdnbHwO3ToUIXeyO12l/sL0TCMctvFxcWMHTuW9957jw4dOvDuu+/yt7/9jdmzZ3t9j4iIehWa+UJERWkGLjOp/asH9YP51AfmUvub6/e0/74j2cz4ZBNH0/IwjFOPt0uIoFv7WP7Qt8WpB8VD3wPVg/rBPNWx7c9Y+H344YcVeqPY2Fg2btzo2U5PTyc6OtqzvWfPHvz8/OjQoQMAt956K6+++up53SMjIx+3+zR/S1eRqKhg0tPzTLt/Xaf2rx7UD+ZTH5hL7W+uC21/p8tNUkoeUz78EYsFWjeuzyWtowmt50tseCDR9QOwWa2e89XHZ6bvgepB/WAes9rearWc9UGYV+/4bdq06azHL7744nNeo0ePHsycOZPMzEwCAgJYtmxZuSGiTZo0ISUlhQMHDngWhk9MTPQmnoiIiMgFKy518tfX11FUUva+3l0DLuLyi+NNTiUiUrG8Xs7BYrFg/GrMg8ViwWKxYLVaPev6nU1MTAyPP/44w4YNw+FwcPPNN9OhQwdGjhzJqFGjSExMZNq0afz5z3/GMAwiIiKYOnXqhX9lIiIiIr+SW1BKfpGDPUeyWbc9BYsFnC6Dg8dzAUhMiGDgpY24qFGYyUlFRCqexTBON4K9vOTk5HLbLpeLgwcP8uqrr/LEE0/Qo0ePSgt4PjTUs25T+1cP6gfzqQ/MpfY318n2d7sNsvJKcLrduFwG3/10nCXfHy53btPYYAL8fDAMg4sahTGoVzOsmqDld9P3QPWgfjBPjR7qGRcXd8q+xo0bExQUxKRJk1i8ePGFJxQRERGpYB8t+5lVW46dsn/k9W2x+1hpFFOPmPqBJiQTETGH1+v4nU5ERARJSUkVlUVERETkgqVlF/F//9lFclo+qZmFxEUGcU23xtisVnxsFmLDA4mLMn8GcBERM1zw5C75+fm8//77tGzZssJDiYiIiHjDMAyKS11s2pPOO//eBYAFSGweQc/EBlzSOvrsFxARqSMueHIXKBsC+sILL1RKMBEREZHTOZyax9INR9iy7wQOpxunyw2Ar4+Vm/q3pF+HWOw+NpNTiohUL14VfitWrDhln91uL7cOn4iIiEhFcjhdnkXUT+QU895Xu8ktLCUtqwgAf18bXVpF0Ti6HvWD/ejaOpoGsaGa0EJE5DQueHIXERERkcpwODWP95fs5uDxUwu4prHBXNI6mmu7NaFJbLAJ6UREaiavCr8jR47w8ssvs3fvXkpLS085vnTp0goPJiIiInVPbkEpb36xg9TMQlrEhdKpZaTnWP1gP7q3izUxnYhIzeVV4fe3v/2N1NRUrrnmGvz9/Ss7k4iIiNRB3207zj//UzZByy2Xt+DqyxqbnEhEpPbwqvDbuXMnc+bMoV27dpWdR0RERGoht2HgcLjZm5xNgK8PDSODsFot2KwWrFYLVouFVVuSARh+TWt6d2xocmIRkdrFq8KvSZMmFBUVVXYWERERqWX2HMnm5XlbKXG4znqeBTCAfp3jVPSJiFQCrwq/8ePH88wzzzB8+HDi4+OxWq3ljl988cWVEk5ERERqrp8PZ/Hcx5sB6JkYS8OIIApLnNQP9qPU4cYwDFxuA7e77HcDgx7tG5icWkSkdvKq8Dt48CD79+9n9OjRpxyzWCzs2rWrwoOJiIhIzeRyuzl4PI8Vm8qGbg67uhX9OmmGcBERM3lV+L322mvcfPPN3HXXXQQEBFR2JhEREanBvtt2nPeX/AxAaJAvvTvoKZ6IiNm8Kvzy8/MZMWIE8fHxlZ1HREREarDDqXks//EoNquFp+/qQlSYP7bfvCIiIiJVz6vCb+DAgSxfvpx77723kuOIiIhITXPsRAGb9qSTX+Rg56Es0rOLuKFnUxIahpgdTUREfuFV4RcXF8fLL7/MsmXLaNKkCT4+5T/tmWeeqZRwIiIiUn2VOlzsPZrDglX7SUrNw89uIyjAh2sua8Kgns3MjiciIr/iVeG3YcMGOnToAMDRo0crNZCIiIhUbyWlLrbuP8EHS36msMQJwICujbj9ypYmJxMRkTPxqvD78MMPKzuHiIiIVGPrd6SQmllIqdPNku8Pe/YP6NqIK7rGExnqb2I6ERE5F68Kv02bNp31uNbxExERqb3yixy8vXgnAHYfK/UC7FzSJpohfRII8rebnE5ERLzhVeF3xx13YLFYMAzDs89isWCxWLBarWzfvr3SAoqIiIg5DMNg3fYUdh7KAuC+G9rSrV2syalERORCeFX4rVixoty2y+Xi4MGDvPrqqzzxxBOVEkxERETM4XaXFXyfrzlAVl4JAO2bhdOqcX2Tk4mIyIXyelbP32rcuDFBQUFMmjSJxYsXV3gwERERqTonsov4euNR3G6DfcdySErJA6BFfCijbupAvQAN6RQRqcm8KvzOJCIigqSkpIrKIiIiIiYoLnXyf1/uZM/RHIL8y/5r0CQmmCdv70Sg3uETEakVLnhyl/z8fN5//31attTUzSIiIjXZZ6sPsOdoDhEh/rzwUA+z44iISCW44MldoGwI6AsvvFApwURERKTiFJU42Z2UxbGMAhxON/uTcygudZFX5CAtq4jGMfV47OaOZscUEZFKckGTuwDY7Xaio6MrPJCIiIhUrBPZRTz11vpy+3xsVvzsVlo3rk/fjg0ZcEkjfGxWkxKKiEhl83pyl3Xr1uF2u+nVqxcAU6ZM4YorrqBbt26VGlBERETOn8vt5h9f7CAjt4SDx3OBspk57x/cjgA/H6wWi8kJRUSkKnn1o72FCxdy3333ceDAAc++nJwcRowYwVdffVVp4UREROTccgtL2bLvBD/+nO75teT7w2z8OZ3DqXl0aB7Bn4d24M+3dCTI366iT0SkDvLqid/s2bOZMGECQ4cO9ex7/vnn6dq1K2+88QbXXHNNpQUUERGRM9uwK5W3vthxxuNjh3WhaWxIFSYSEZHqyKvCLzk5+bRDOrt3786UKVO8vtnixYt58803cTqd3HPPPdx5553ljh84cIAJEyaQk5NDVFQUM2bMIDQ01Ovri4iI1CWb96R7ir6Hb0wkKsy/3HF/XxvR9QPNiCYiItWMV0M9GzduzOrVq0/Zv3btWho0aODVjVJTU3n55Zf5+OOPWbhwIXPnzmXfvn2e44Zh8OCDDzJy5EgWLVpEmzZtmD17tpdfhoiISN2QX+Rg85501m9PYcHq/QA8OiSRLq2iaBwTXO6Xij4RETnJqyd+f/rTnxg3bhw7d+4kMTERgO3bt7No0SL+/ve/e3WjdevW0a1bN8LCwgAYOHAgS5Ys4ZFHHgFgx44dBAYG0qdPHwAeeOABcnNzz/sLEhERqc3mrtjL2u0pnu02TerT+aIoExOJiEhN4FXh94c//AFfX18++OADvvrqK+x2OwkJCbz88stceeWVXt0oLS2NqKj//cMUHR3Ntm3bPNuHDx8mMjKSMWPGsGvXLhISEhg/fvx5fTEREfXO6/zKEBUVbHaEOk3tXz2oH8ynPjBXRbb/V+sO8uPuNNKyCjmRXUReoYMW8aE8eVdXAvx8CAnyxaZlGMrRn3/zqQ+qB/WDeapj23tV+AFce+21XHnllfj6+gJw7NgxGjZs6PWN3G43ll/NImYYRrltp9PJhg0b+Oijj0hMTOSVV15h+vTpTJ8+3et7ZGTk43Yb5z6xkkRFBZOenmfa/es6tX/1oH4wn/rAXL+3/bfsPUGp04XD6Sa3sJT535QN52zdOIy8QgcA8VH1sGPgLHGQWeKokNy1hf78m099UD2oH8xjVttbrZazPgjzqvBLT0/nscceo2vXrvzlL38B4KabbqJFixa8+uqrhIeHn/MasbGxbNy4sdw1f70AfFRUFE2aNPEMJb3++usZNWqUN/FERERqrFKHi+U/HiUtq5A1245j/Obnl352Gw8MbkfHFpHkFZZyJC2fZg00S6eIiJwfr8aGPPvss1gsFoYMGeLZ99FHH+F2u71+ItejRw/Wr19PZmYmRUVFLFu2zPM+H0Dnzp3JzMxk9+7dAKxcuZJ27dqdz9ciIiJS4+w4mMmCVfv5blsKhgFdW0cz+s6LmX5/N15/vA9v/KUPHVtEAhAc6EvbpuEE+Hk9YEdERATw8onf+vXrmTNnDk2bNvXsa968OePHj+fee+/16kYxMTE8/vjjDBs2DIfDwc0330yHDh0YOXIko0aNIjExkddff51x48ZRVFREbGwszz///IV8TSIiItWSy+0mLasIl8ug2OGipNTFTwcyAJjxaE9CAn1NTigiIrWVV4WfxWKhqKjolP0ulwuHw/t3C2644QZuuOGGcvvefvttz8cdO3ZkwYIFXl9PRESkJlm45iD/Xp90yv6QQDvBAXYTEomISF3hVeHXq1cvpk6dyowZMzwTuhw/fpzp06fTs2fPSg0oIiJS06VnF7F47SG+++k4NquF+we1w8/Xhp/dhr+vjfAQ/3ITnomIiFQ0rwq/MWPGMHz4cK644grPRC6ZmZm0adOGF198sVIDioiI1HQLVu3nh91p+PvaeHhIIu2anntSNBERkYrkVeEXERHB559/zrp169i7dy8+Pj40b96cHj166CeUIiIi51BQ7CAy1J8pI7th99GaeyIiUvW8nhbMZrPRu3dvevfuDZStw7dq1Srmzp3LW2+9VWkBRUREaqIVPx5l6/4TOBxuklLzaB4XqqJPRERMc97zQaekpDB//nw+/fRTUlJSiIyMrIxcIiIiNdrXPxyhsMRJw8ggWsSF0rtjQ7MjiYhIHeZV4WcYBt988w3z5s1jzZo1uN1uWrZsyaOPPnrKLJ0iIiICJQ4XXVpFcc/Vrc2OIiIicvbC7/jx456ne6mpqURERHDLLbcwb948ZsyYQYsWLaoqp4iISI2QlJLLtxuPkFNQip/dZnYcERER4CyF3/333893331HaGgoV1xxBddccw3dunXDarUyb968qswoIiJSIzhdbv428zsKip342q1c1CjM7EgiIiLAWQq/1atX06xZM/74xz/Sq1cvGjRoUJW5REREqiWH08WqzcdIzSrEggW7j5UTucVs2XsCp8sNwJ0DLuKKLvEmJxUREfmfMxZ+8+fP5/PPP+fFF1/k73//O+3bt+faa69l4MCBVZlPRESkWvlxTzqfrNhbbl90/QBaxofSJDaYRrEhdErQOn0iIlK9nLHwS0xMJDExkaeffprly5ezcOFCXnzxRZ5//nkAVq5cSXx8PP7+/lUWVkRExGxZeSUADO3XnKAAO707NCi3pm1UVDDp6XlmxRMRETmtcy4oZLfbueaaa/jHP/7B6tWr+etf/0qLFi2YMWMGvXv3Ztq0aVWRU0REpFrIK3Bg97Fy9WWN6dOxYbmiT0REpLo6r5VkIyMjGTFiBIsXL2b+/Plcf/31fPHFF5WVTUREpFpJTs9n9dZkGkYGqeATEZEa5bwXcD/p5FDQMWPGVGQeERGRaqm41Mmcr/dQVOLi/kHtzI4jIiJyXs7rid/p2O32isghIiJSra3Zepzdh7O5vkcTYsMDzY4jIiJyXn534SciIlIX5Bc5sAA39k4wO4qIiMh5U+EnIiLihVKnC7vdqnf7RESkRvLqHb/S0lL+7//+j+uuu44mTZowefJkPv/8czp27MhLL71EREREZecUERGpUl+uO8SyH44AYBgGJQ43/r42k1OJiIhcGK+e+D3//PPMmTOHwsJCVq1axbx583j44YdxOBxazkFERGqlHQczsftYuaRNNJe1jaFvx4bcdkULs2OJiIhcEK+e+C1ZsoQZM2bQpk0bPvroIy677DJGjBhBz549uffeeys5ooiISNXLK3KQ0CCEu69qZXYUERGR382rJ355eXk0adIEgLVr19KrVy8A6tWrR2lpaeWlExERqWIrfjzKv1bs5diJAoIDNXO1iIjUDl498UtISGDNmjVER0eTkpJCnz59AFiwYAEtWmjYi4iI1EwOp4v3vvqZm/omEB7iz5qtx5jz9R7P8U4to0xMJyIiUnG8Kvwee+wxHn30UZxOJ9deey3Nmzdn+vTpzJkzhzfeeKOyM4qIiFSK7QcyWb8jhT1HsrmpXwLvfrWbJrHB/Om6NgT6+RAe4m92RBERkQrhVeHXr18/vv32W1JSUmjTpg0A119/PbfffrtnCKiIiEhNY7eXvfGQkVvM7EU7Abi4ZSTxUfXMjCUiIlLhvCr8rrjiCj799FNP0QfQvn170tLS6N69O+vXr6+0gCIiIpXB6XLz6aoDADz4h/bERwXh62MjPMTP5GQiIiIV74yF3+rVq/npp58ASE5OZvbs2QQGBpY759ChQ7hcrspNKCIiUgkycopJSs3jxj4JXNI62uw4IiIileqMhV98fDxTp07FMAwAli5dis32v4VrLRYLQUFBjBs3rvJTioiIVKAjafls238CgGaxwSanERERqXxnLPyaN2/O0qVLAbj77ruZNWsWoaGhVRZMRESkMhxOzWPiuz94tqPDA89ytoiISO3g1Tp+H374YYUUfYsXL+baa6/lqquuYs6cOWc8b9WqVfTv3/93309EROTXTmQXeYq++wa1ZcK9lxAdFmByKhERkcrn1eQuBw4c4Nlnn2Xz5s04HI5Tjm/fvv2c10hNTeXll1/ms88+w9fXl9tuu43LLrvslHUAT5w4wXPPPedlfBERkVOVOFws33iE5BMFZOYUczS9AIfLjcPpBqBn+1guaxODxWIxOamIiEjV8KrwmzRpEseOHePPf/4zISEhF3SjdevW0a1bN8LCwgAYOHAgS5Ys4ZFHHil33rhx43jkkUd46aWXLug+IiJSt326ej//Xp/k2W4UXY9WjcOICQ/Ex2YlPNiPvp0aqugTEZE6xavC76effuLdd9+lY8eOF3yjtLQ0oqKiPNvR0dFs27at3DkffPABbdu2/V33ERGRuueH3Wls3ptOdl4Juw9nA3DHlS3p2jqasHpankFERMSrwi8sLAxfX9/fdSO3213up6uGYZTb3rNnD8uWLeO9994jJSXlgu4REWH+grtRUZodzkxq/+pB/WC+utYHC/6xnsJiJ41jgunZoSE39E6gXUKEaXnqWvtXN2p/86kPqgf1g3mqY9t7Vfg98MADTJ06lcmTJ9OkSROsVq/mhCknNjaWjRs3erbT09OJjv7fuklLliwhPT2dm266CYfDQVpaGnfccQcff/yx1/fIyMjH7TbOO1tFiYoKJj09z7T713Vq/+pB/WC+utYHi9YeJD2riEE9m/KH3gme/Wa1QV1r/+pG7W8+9UH1oH4wj1ltb7VazvogzKvC79133+XIkSNce+21WCyWUwo/byZ36dGjBzNnziQzM5OAgACWLVvGM8884zk+atQoRo0aBcDRo0cZNmzYeRV9IiJSu+Tkl/Dv9Ulk5BZTWOwkK68E129+uOdwucktKAWgvYlP+ERERKo7rwq/++6773ffKCYmhscff5xhw4bhcDi4+eab6dChAyNHjmTUqFEkJib+7nuIiEjNlZVXwp4j2Thdblxug52HMtmwK434qCAC/e3ER9cjwNd2yuf5+FgZ1LMZ9YP1Lp+IiMiZWAzDMG9sZAXTUM+6Te1fPagfzFeT+mB/cg4LvztIRk4xKZmFpxxvEBHIlJHdTEh24WpS+9dGan/zqQ+qB/WDeWrcUM/x48czevRogoKCGD9+/BkvYLFYmDx58u9LKSIidc6vl11oGR9KQsNYEhqG0D4hAh+rBR+blUB/rwamiIiIyDmc8V/UQ4cO4XK5PB+LiIicL6fLzTebk0nPKsLhcmMYBm532czO3+9KpUVcKA/d2F5LLoiIiFSyMxZ+H3744Wk/FhER8UZOfglTPvyREznFAIQG+WKxlA1FsWAhIsSfQb2aqugTERGpAl6PoTl27Bhz5sxh7969+Pj40LJlS2655Rbi4uIqM5+IiNQQJaUulm08wuY96RSWOEnLKgKgSUwwo++6GD/7qROziIiISNXwqvDbtWsXd911FyEhIXTo0AG3280XX3zBRx99xJw5c2jdunVl5xQRkWrA6XJT6nB5tvcczWHdT8c5kl5A6q8mZ7m0TTTtm4XTICKI/hfHYbFYzIgrIiIiv/Cq8Hvuuefo06cPzz//PHa7HQCHw8Ho0aN54YUXeOeddyo1pIiImCu/yEFBsYMXP9lCRm7xKcdbxIfSunFDIkL86ZnYQEsriIiIVDNeFX5btmxh/vz5nqIPwG63c//993PbbbdVWjgREal8hmHgdBkUFDv4fmcq3207Tlp2EU6nu+z4b85v3TiMTi2jPNvtmoUTFxlUhYlFRETkfHlV+IWEhFBQUHDK/vz8fHx8NNW2iEhNYBgGeUUOdh7KZOWmZDJyiikpdVHicOH6zRqo3dvFEBEawMkBmm7DwNduIy4yiM4tIzV0U0REpIbxqmrr168fkydP5pVXXqFx48ZA2RIPU6ZMoW/fvpUaUEREzs1tGOw8lEl+oQNf/wxOZBZS6igr6kocLo6m5XMkLZ+CYqfnc4L8fejRPhY/Xxv+vjb87Db8fX1o27Q+4SH+Jn41IiIiUtG8Kvz+8pe/MHz4cAYOHEj9+vUByMrKomPHjjz99NOVGlBERMrkFpSy/1gOTpeBw+miqMSF2zDAgJ2HMtm6P+OUz7FaLPj5WvH39aFpgxBaNQqjYWQQCQ1DflleQU/uRERE6gKvCr+wsDA+/fRT1qxZw969e/Hz86NFixZ07969svOJiNRJP/6cxrrtKeQVOjz79iXnnPPzxt/TlYTG4eTnFeHrY8PHZlFxJyIiImcv/AoLC/nvf/+Lr68vnTt3pm/fvhraKSJSAQzD4ODxPDbvTWfjz+kYhoHbbeByG2TllQDg72ujWYMQTtZtbZvWJ6FhKJe2icbHZsXPbsPuY8ViAQtg9ynbjgwLwHA4z3xzERERqXPOWPjt3r2bESNGcOLECQCio6OZNWsWHTp0qLJwIiI1WfKJArbsTaew2Inrl6KurLhzs/doDscz/rfuXYfmEQT527HZLFiAoAA713ZrQr0A+5lvICIiIuKlMxZ+L730Eo0aNWLWrFlYrVZeeuklJk2axKefflqV+UREaoR9R3NIzynC5Sor7L5cl1RuvTs/Xxs+VgtWqwWb1YLNaiUuMogR17elQUQgvnabielFRESktjtj4bdlyxY++OAD2rRpA8Czzz7LwIEDKSwsJDAwsMoCiohUV27DYPnGo2zak86eI9mnHG/TpD53DLiIBhGBWPWencj/t3fn0VHX9/7Hn7NmXwhkgSQECCGsURAvyNYCgqLVCkivFgFtpbfucu61Pd5q1WNv8VcX2otX29vDr7VW6tofFQWxVKUiiEKLSUgCEkLYQhKyQPbJzHx/f0SnDYKGbT4zk9fjHA7MfCeT97zfmbx5z/f7/XxFRMSg0w5+LS0t9OvXL3A7Ozsbh8NBY2OjBj8R6VVqG9tY834Fh2qb8fr8gb169U0dWJ9d/i4vK4lrJg8ivU8sDrsNp8NOQqxLC6uIiIhISDjt4Of3+7Hb7d0f7HTi8/kueFAiIsHk8/v5+55jVDe0drv/aH0r9Sc6KK1sALquezdiUApOuw2Ho+twzYRYF9dOHozLaT/VU4uIiIiEhB5dzkFEJJx9eqiRD4qqaGjy0Njc0XXtO6C13YvX5+92yYSTRbkd5GUlMWpQCtdOGRyskEVERETOqy8d/H73u98RExMTuO3z+Vi9ejVJSUndHvf973//wkQnInIGdh9o4NVN5Rw73o7/sxU08H5TlAAAIABJREFUW9u9fHY0JgP6xZGWHIPD3nX4pd1uIy7GhcNmIznBzcSRGSTGubs9p66DJyIiIpHgtIPfgAEDWLt2bbf7+vXrx4YNG7rdZ7PZNPiJSNCUHznOxu2H8Pr8WBZdA55lcbi2JbCKZma/OPKyk7HbwG6zkRDnpmBIX3IyEgxHLyIiImLGaQe/d955J5hxiIicUnNbJ//3zVIqqk7Q0t6J19e1/y7K5SA1ORqbzYbdZsPltJOVGs+C6bmMGdLXcNQiIiIioUXn+IlISLAsi+KKetZ/WBlYKROg/MgJvD4/WanxjM9Pw263MX54KkMzk3QIpoiIiEgPafATkaDbXlZD6YEGLL+F34L9R09Q3dBGh6dr1eD87OTAY3MHJNInIYrvfmMEDrtWzhQRERE5Gxr8RCSoCsuP8cyaYpwOO7FRDmz2rkM1B/SN5ZL8NEbk9GFw/0TTYYqIiIhEFA1+InJB+P0WW4qP0ubxggUWUN3Qyrt/OwzAjTOHMn1cltkgRURERHoJDX4icko+v59Orx+/38L32SGZXf/2s+/ICfYePk5bh5eOzs8e4/PjsywcDjttbZ3sr27C0+k/5XP/8NtjyR/YJ8ivSERERKT30uAnIgGWZdHu8bGrop5n1hR/5eNTEqOIcjlw2G3Y7TYcdjvRUU6w2RiZk8LwnD5cNiodm82G7bNLK8RE6deOiIiISLDpf2AivZDX52d7WQ3FFfXUHW+noamD1g4vbR1efP5/LKk5Nq8f+dnJnw11XcPd5/8ePrAPKYnRX3ju1NQEamubgvlyREREROQraPAT6QV27j1G8b46fH4Lr8/P3/Yco63DC8DQrCQG9U8gLtpFTJST2GgnackxZKfFk9YnRpdMEBEREYkAQR381q5dy7PPPovX62XJkiUsXLiw2/aNGzeycuVKLMsiKyuL5cuXk5SUFMwQRSLKK+/tZeP2Q3R6u861S4pz43TYSIh1kZeVxOIr8k+5105EREREIkvQBr/q6mpWrFjBH//4R9xuNzfccAMTJkxg6NChADQ3N/Pwww/z2muvkZ6ezi9+8QtWrlzJAw88EKwQRcKW19e1EIvPbwUWY1n/YSUbdxwirU8MI3L6MGfCQNL6xJoOVUREREQMCNrgt2XLFiZOnEhycteFma+44greeust7rzzTgA6Ozt56KGHSE9PByA/P5+1a9cGKzyRsFXd0MqPV30U2Kt3siVX5DNiUEqQoxIRERGRUBK0wa+mpobU1NTA7bS0NAoLCwO3+/Tpw6xZswBob2/nf//3f1m0aNEZfY++fePPT7DnIDU1wXQIvVpvy/+hmib+zwt/p9Pr59tXDCcu2onDYcdht+F22Zk4uj+x0a6gx9Xb6hCKVAOzlH+zlH/zVIPQoDqYE4q5D9rg5/f7uy0SYVnWKReNaGpq4o477mD48OHMnTv3jL5HXV0z/n9akTDYtJqhWb0p/y3tnTy/YTcfldYAMGfiQGZe3P8L76mWpnZamtqDGltvqkOoUg3MUv7NUv7NUw1Cg+pgjqnc2+22L90RFrTBLyMjg+3btwdu19bWkpaW1u0xNTU1fPe732XixIn853/+Z7BCEwlZh2ub2fDxQcoqG4Cu6+DZbFDd0BZ4zL0LLqIgt6+pEEVEREQkDARt8Js0aRIrV66kvr6emJgY3n77bR599NHAdp/Px/e//33mzJnD7bffHqywRIzzW10LsmwtPso7fzuMx+sLbKuqawUgNzORtOQYLMCyICcjgdTkGOZOHYLdrsstiIiIiMiXC9rgl56ezrJly1i8eDGdnZ1cf/31FBQUsHTpUu6++26OHj1KSUkJPp+PDRs2ADB69Gj+67/+K1ghilxwVXUtdHr9WBY0NHXw9B+L8FvdD08e3D+Rvkldl1gYNSiFyWP6k5MReseJi4iIiEj4COp1/K655hquueaabvf9+te/BmDMmDGUlZUFMxyRoCnZX8+vXt9FU2vnF7aNH55GSkIUfsvi6ok5JMVHGYhQRERERCJZUAc/kd7meHMHG3cc4s2tlQDM/9oQ+veNw2YDm81GXLSTvKxkw1GKiIiISKTT4CdynmwtPkpJZT3tHT4S493ERbt4a1slXl/XoZyXj8/i6ssGmQ1SRERERHolDX4iJ/FbFlXHWjh2vB2vz8Ln99PW4aX+RAc2G/j8XYux/PPfXp+f9wurAIiLdtLa4QW6FmK5eGg/brtuNC6n3eTLEhEREZFeTIOf9Br7j56gscmDx+vD0+nH6/v8j0W7x0tNYxu1jW0crWulpd172udx2G3YP/vjsHX97XTYSEuO4YaZeVyc1y+Ir0pERERE5Ktp8JOI1dreSWllAz6/RXV9K//v/YovfXxctJOMlFjGDkslOzWe/v1iSYx147DbcDjsxEY5SYh1feEi6SIiIiIioU6Dn0Sc4oo6fru+jBMtnXh9/m7bbv3GCHLSE3C5HLgcdpwOG06HHafDrkMxRURERCRiafCTsNbp9XPseBten0Vju5fiPbU8//Zu7HYbM8Zlckl+KrHRLgBi3A5SEqMNRywiIiIiEnwa/CRsbS+r4dk1xVin2HblhIF8a/rQoMckIiIiIhKKNPhJSLIsi31VJzjR7OF4q4dO7z8O2SypqKexxUPl0Saga9XMyWMySEmJo6WpnbSUWNKSY0yFLiIiIiIScjT4SUgqP3KCnz6/47TbU5OjmTEuk1mXZpPeJ7brvtQEamubghWiiIiIiEjY0OAnxliWRWuHF5+v61p4Pp+fiqNNvPzOXupOtANw1/wxZPaLC5ynB2C30e22iIiIiIh8OQ1+EnTHGtvYXFTFX3YcOu318sbnpzKgXxwXD+2nyyeIiIiIiJwjDX5ywb387l72HjpOm8dLp9dPTUMbAFEuB6MHp3BxXr+ui6HbbTjtdvolR5OXlWw4ahERERGRyKHBT85JW4eX9wur2H/0BJ5OPz6fv+uwTb/F8RYP1fWt+PwWCbEuhmUl43Layc9OJn9gMpeNytDePBERERGRINDgJ2el0+tnc1EVr7y7l3aPj4RYF4lxbhx2Gw67HYfDRp+EKIZmJpIYF8WcCQOJidKPm4iIiIiICfqfuPTY3/bUsq2kmk6vn6J9dfj8Fv37xjK1YABXThhoOjwRERERETkNDX7ypT4uq6Gw/BiHalsC183LTotnYHoCuZmJ3DgzT4drioiIiIiEOA1+clpPvbyT4n31xEQ5GZSRwJQx/Zk8JoP8gX1MhyYiIiIiImdAg58EeH1+1rxfQVOrhxMtHor31TNqcAp3zy/A5bSbDk9ERERERM6SBr9exLIsDtW2UFbZgNfnx+u36PT6OdHSwfFmD5+U1wUem5Ycw6CMBBbNHqahT0REREQkzGnw6wU2F1bx10+OUHu8jePNni9sj4lykhjnZmB6PFMLBvC1iwfgdGjYExERERGJFBr8IkRru5fmNg9lBxpp6/AG7rcsePfvh2ht95KTkcDEkelMHtOf1KQYHA4bdrsNuxZnERERERGJaBr8wohlWez89BgfllRTsr8eywK/ZeG3LDyd/tN+XUyUk/lfG8KMcVlBjFZEREREREKFBr8wsPfQcdZvq6Si6gSNzR5cTjtOh50JI9NxOezY7eByOkhLjiHa7WBoVhJRLkfg66NcDux27dUTEREREemtNPiFmOqGVv62p5aahja8Xj+dPj/7jpzgRKuH7LR4Ls5L5duX5+kcPBERERER6TENfiGi/PBxivbV8foH+wGIj3ER5XLgdNqJdjuYfWkul4/PNhukiIiIiIiEJQ1+F1B1fSvbSqrx+i38n/3x+v2caPHg6fRT09iG3QY+v0VVXSsAfROjuG7qECaP6W84ehERERERiRQa/C6g4op61myuwGYDh/0fK2jGRDmJi3bhsNtIjHURE+1iUEYCl4/PZnD/RNNhi4iIiIhIhAnq4Ld27VqeffZZvF4vS5YsYeHChd22l5aW8qMf/YiWlhbGjx/PI488gtMZvrPpzEuymD4uU5dLEBERERERo4K2Qkh1dTUrVqxg9erVrFmzhpdeeom9e/d2e8x9993Hj3/8YzZs2IBlWbz88svBCu+C0dAnIiIiIiKmBW3w27JlCxMnTiQ5OZnY2FiuuOIK3nrrrcD2w4cP097ezsUXXwzAvHnzum0XERERERGRsxO0wa+mpobU1NTA7bS0NKqrq0+7PTU1tdt2EREREREROTtBO4HO7/dj+6fDHi3L6nb7q7b3RN++8ece6DlKTU0wHUKvpvyHBtXBPNXALOXfLOXfPNUgNKgO5oRi7oM2+GVkZLB9+/bA7draWtLS0rptr62tDdw+duxYt+09UVfXjN9vnXuwZyk1NYHa2iZj37+3U/5Dg+pgnmpglvJvlvJvnmoQGlQHc0zl3m63femOsKAd6jlp0iS2bt1KfX09bW1tvP3220ybNi2wPTMzk6ioKHbs2AHAn/70p27bRURERERE5OwEbfBLT09n2bJlLF68mOuuu45vfOMbFBQUsHTpUoqKigB44oknWL58OVdeeSWtra0sXrw4WOGJiIiIiIhELJtlWeaOjTzPdKhn76b8hwbVwTzVwCzl3yzl3zzVIDSoDuaE6qGe4Xt19FOw281fMy8UYujNlP/QoDqYpxqYpfybpfybpxqEBtXBHBO5/6rvGVF7/EREREREROSLgnaOn4iIiIiIiJihwU9ERERERCTCafATERERERGJcBr8REREREREIpwGPxERERERkQinwU9ERERERCTCafATERERERGJcBr8REREREREIpwGPxERERERkQinwU/CjmVZpkMQERGD1AdERM6cBr+z5Pf7TYfQax0/fhxQ4xcRs9QHzFEfEJFQEG59QIPfGSgrK+Pxxx8H1GxM+Pjjj5k9ezaPPvooADabzXBEvc+HH37Ixo0baW5uNh1Kr7V//3527NgB6PeQCeoDZqkPhAb1AvPUC8wJ5z7gNB1AONmzZw+rVq1i3rx55Obm4vP5cDgcpsOKeBUVFfzP//wPR48eJTY2lqlTpwJdn7LY7frsIhg8Hg8PPvgg5eXl5OXlsX79eubOncuUKVNUhyBbuXIlffr0IT8/n/j4eNPh9DrqA2aoD4QG9YLQoV5gTjj3Ab1Dz8Dx48cZMGAA9913H0DYFDmctbW18ZOf/ITc3Fx+//vfM3/+fCoqKgDUYIKora0Nu93OH/7wB5YvX86UKVN44IEHANUhWCzLoqysjG3btlFVVcU777xjOqReSX0g+NQHQod6gXnqBeaFcx/Qu/Q0Nm3aRFlZGT6fDwCfz0dVVRWrVq2ivr6eP//5z4HHhttu3nCwadMmSktLiYmJYdWqVdx2221A16csOTk5QPgdVx1uysrKqKmpAboOKSkuLsblcmFZFnPnziU1NZUnnngCUC0ulM9/D3m9Xmw2G0eOHOH6669nwoQJ7Nq1iwMHDgD6HXShqA+YpT4QGtQLzFMvMCfS+oDj4Ycffth0EKGkurqaW265heLiYoqKiigrKyM3N5eEhATeeOMNpk+fztixY7n33nvZuXMnkyZNIjY21nTYEePz/BcVFVFcXExJSQnDhg0L5Li0tJRdu3Yxa9YsndtxAR05coQf/vCH9O/fn6FDh5KRkcErr7xCe3s7Y8eOBeCiiy7iiSeeYP78+URHRxuOOLKc6vfQyJEjcblcTJ06laSkJD755BOam5spKCjQe+E8Ux8wS30gdKgXmKVeYE6k9gHt8TvJvn37+Jd/+Reee+457rnnHjweD0899RTQNck3Njby3nvv4Xa76ezsJCUlhc7OTsNRR47P8/+73/0ukP+f/exnge0jRowgOTmZxsZGg1FGvvfff5/du3ezY8cOysrKAFi2bBl/+MMfaG9vByAvL4/x48cHTi6X8+fk30Pt7e0sX76coUOHkpCQQH5+PsOGDaO8vJzCwkIgPD5pDBfqA2apD4QO9QKz1AvMidQ+oMHvJIWFhXz66acA5OTksGTJEoqKivj4449pa2vj5ptvJj09ndWrV/PBBx9QV1eHy+UyHHXkOFX+S0pK2Lp1KwBOp5Pq6mqcTq1LdCE1NTXxH//xH7S3t1NcXExHRwfTpk1j2LBhgfM5Ojs7aW1tZcSIEYajjTwnvw9uueUWdu3aFXgfAEybNg23282f//xnPB6PPuk9j9QHzFIfCB3qBWapF5gTqX1Ah3p+xrIsbDYb2dnZvPTSS4wePZqMjAwSEhIAePXVV/n2t7/NPffcw6RJk0hJSQmsphQVFaU32jn6svz7fD7Wr1/PVVddRU5ODo8//jjx8fGMGTPGdNgRq76+nrlz51JTU0NhYSFpaWlkZGQwbdo0Xn75ZXbs2MHKlSvJzc1l9uzZOBwOvQfOwec//1/1Pnjrrbe46qqrAEhMTKShoYG4uDiGDx8eVieXh5qe5B/UBy6Unv78qw9cWJ/n/5+pFwSXeoE5vaUP9MrBr6SkhHfffZfExESSkpIC99tsNlwuFw0NDWzevJlZs2ZhWRadnZ1UVFQwd+7cwK5ch8NBQUEB0dHRIV/kUHOm+ff7/ZSVlTFq1CgSExMZNmwYmZmZpKenG3wV4e1UNfg813a7nSFDhgAwaNAg3n//fTweDwMHDiQ5OZkZM2aQk5PD9OnTufHGG3E6nXoPnIWSkhK2bNnCgAEDup0X82Xvg9LSUi666CLi4uKw2Wzk5eUxevRoNfqzcKb5Vx84v87m51994Pw7XR18Pp96QZBUV1dTWFhIdnZ2IH+f/61ecGGdae4joQ/0qsHP7/fz2GOP8fvf/574+HjWrFnDwYMHufTSSwPFczqd9OvXjw0bNtDa2sqYMWM4cuQIhYWFXHfddUB4LdsaSs42/4cPH6aoqIjrr78egOzsbDX7s/RVNTj5MAW3243dbmfz5s0kJCQwZMgQoqOjSUtLUw3OkmVZrFixgl//+te0t7ezadMmTpw4wciRI+no6MDpdH7p+2DevHlfaFDSc2ebf/WB8+Ncf/7VB86Pr6qDekHwPP3002zfvp2LL76Y+Ph4/H5/YPBWL7iwzjT3kdAHetUB8o2NjRw9epTnnnuOhIQEioqKWLBgAbNmzSIvLw+A3/72t7hcLm677TYefPBB9uzZw5YtW1i4cCFw6kMhpGeUf/N6UoPf/OY32Gw2br75ZgCmT5/Oli1biIqKMhh55PD7/TQ0NPDLX/6SzMxMtmzZwr333svMmTPp06cPoPfBhaT8m6X8h4ae1EG94MLy+/3U1tayadMm0tLS2LhxIwsXLsRutweuiaj3woXRm3MfsXv8Pi9IeXk5zc3NJCUlUVVVxZNPPsnSpUsDJ4W/9957VFZWMm3aNObPn4/X6+Vf//VfGTFiBDNmzCAuLo6bb76ZadOmAfpUpaeUf/POpgbXX389Xq+Xm266icTExMAnXxMnTmTw4MGGX1H42rJlC7t37yYlJQWv18svfvELbrjhBmJiYsjOzmb37t1s2LCBq6++mmuvvVbvg/NM+TdL+Q8NZ1KHb37zm+oFF8DnNUhKSiIuLo7y8nL8fj/jxo1j3759JCcnk56eztGjR1m0aBGdnZ16L5wnyn0XmxXB6752dnayZMkSrrnmGubNm0dUVBS33347Pp+PxYsXs27dOsaNG8djjz3GqlWr6NOnD9nZ2UD4TvKhRPk3TzUw6/jx49x33320traSnZ1NU1MTDz/8ME8++SROp5NHH30UgObmZubMmcMvf/lLEhMTVYPzRPk3S/kPDaqDeSfXoKWlhbvuuovo6GjcbjcAq1atIiUlhe985zu43W6Ki4sZPXo0oBqcC+W+u4jd4wdd0/1zzz1HfHw86enp9O/fn6997WvU1dWxdetWhgwZwqJFizhw4AB+v58pU6YABBa4kHOj/JunGpi1bds26urq+O///m8uvfRSPv30UxobG/nmN7/JM888w+TJk0lJScHtdnPo0CG8Xi+TJ08GVIPzQfk3S/kPDaqDeSfXYO/evZw4cYIpU6YQHx9PfHw8TU1NlJaW4na7ycnJIS0tDVANzpVy311EvZqPPvqIqqqqwO0jR46wbNky4uPj2blzJ/X19cTHx/O9732Pe+65h+9973tA13LFl1xySeDrIq3IwaL8m6camPfRRx9x+PBhoGvFsKamJgCSkpKoqKjAZrMxaNAgZs+ezY9+9KPA19XV1TFq1KjAbdXg7Cj/Zin/oUF1MO/LalBeXh44V9Lr9QIwZcoU+vbty6ZNm6irqws8j2pw5pT704uIxV327dvHXXfdRXp6Ona7nTlz5nDttdeSnZ3NhAkT+Otf/8q6desYNmxY4FOs5cuXk5KSQlFREYMGDSIjIyPiducGi/Jvnmpg3j/XwGazcfXVVzNjxgymTp1Ka2srUVFRNDY2BhZO+MEPfsCtt97KAw88QGlpKRkZGWRmZqoGZ0n5N0v5Dw2qg3lnWgOn04llWSQmJjJ27FgqKioChyDKmVHuv1pYn+Pn8Xhwu9288sorNDc3c8stt/DOO+/w3nvvkZmZyb/9278FHvvTn/6U+Ph45s2bR1ZWFkeOHKG6uprGxkamT59u8FWEL+XfPNXAvFPV4C9/+QubN2+mb9++3HnnnQCUl5dz//33s3r16sDCOpZlsW/fPmpqarjssstMvoywpfybpfyHBtXBvLOtQWtrK7GxsYajD2/Kfc+F5Tl+Xq+XFStWsHbtWtxuN+vWrcOyLKZNm8aAAQOIjY3l9ddfZ9SoUfTt2xeA5ORkNm3ahMvlIj8/n8TERPr376/Vqc6C8m+eamDel9UgMzOT6Oho3njjDUaOHEnfvn158cUXycjIYPLkyTz11FOsWbOGyy67jIyMjMAiCtJzyr9Zyn9oUB3MO9cavPnmm0yePPkL106Ur6bcn7mwO3i1vr6eu+++m9bWVr7+9a/z+uuvM3z4cKqqqjh06BBRUVEMHz6csWPH8sYbbwS+bsyYMeTk5OBwOAjjnZzGKf/mqQbm9bQG48aN48033wRg586dlJSUsHDhQurr67n//vuJj483/ErCk/JvlvIfGlQH885XDXrbXqfzQbk/O2F3jl99fT319fU888wzAFRUVPDJJ58wfPhw1qxZw5133klycjIZGRlUVFTQ0dGB0+nE4XBw22239aqp/kJQ/s1TDcw7kxqUl5cD4PP5aG5u5pFHHmH48OEmww97yr9Zyn9oUB3MUw3MUe7PTtjt8UtISODKK6+kpqYGgMzMTFJTUykoKGDPnj1s2rQJh8OB2+3G4/EQFRWFw+EA0H94zwPl3zzVwLwzqYHP5wPgkUce4aWXXuq1zeZ8Uv7NUv5Dg+pgnmpgjnJ/dsLuHL+4uDiGDBkSOG/pZz/7GVOnTmXOnDl4vV4ef/xxDh06xEsvvcT8+fPJy8vT6lTnkfJvnmpg3pnUYN68eeTl5ZGYmGg46sih/Jul/IcG1cE81cAc5f7shPWqngcPHmTRokW88cYbxMfHU1lZSWVlJc3NzRQUFJCVlWU6xIim/JunGpinGpil/Jul/IcG1cE81cAc5b7nwu4cv3928OBBZs6cSW1tLffccw9JSUncf//9pKammg6tV1D+zVMNzFMNzFL+zVL+Q4PqYJ5qYI5y33NhPfiVlZXxwgsvUFRUxIIFC1iwYIHpkHoV5d881cA81cAs5d8s5T80qA7mqQbmKPc9F9aHer722mscPXqUpUuX4na7TYfT6yj/5qkG5qkGZin/Zin/oUF1ME81MEe577mwHvy0YIVZyr95qoF5qoFZyr9Zyn9oUB3MUw3MUe57LqwHPxEREREREflqYXcdPxERERERETkzGvxEREREREQinAY/ERERERGRCBfWl3MQERE532bMmMHhw4cDt10uF+np6cyePZs77riD+Pj4Hj3P3r17OXToEF//+tcvUKQiIiI9pz1+IiIiJ1m6dCmbN29m8+bNrFu3jn//93/nzTff5NZbb8Xj8fToOW6//XaKiooucKQiIiI9oz1+IiIiJ4mNjSU1NTVwe+DAgeTk5DB//nxee+01brzxxq98Di2aLSIioUR7/ERERHpg1KhRXHLJJaxbtw6ADRs2MH/+fAoKCrjooou44YYbKCwsBGDRokUcOHCAp59+mhkzZgDg8Xh47LHHmDJlCuPGjeOmm25i586dxl6PiIj0Lhr8REREemjYsGHs2bOHwsJC7r33XubNm8e6det4/vnnAXjwwQcBWLlyJZmZmXznO9/h1VdfBeAHP/gBH3/8MT//+c957bXXmDhxIosXL6aiosLY6xERkd5Dg5+IiEgPJSYm0tzcjMvl4qGHHmLhwoVkZWVRUFDAggUL2LNnDwDJyck4HA5iY2NJSUmhsrKS9evX89hjjzF+/HgGDx7MnXfeySWXXMJvfvMbw69KRER6A53jJyIi0kMtLS0kJCQwYsQIEhIS+NWvfsXevXuprKyktLQUv99/yq8rKSkB4Fvf+la3+z0eT48XixERETkXGvxERER6aNeuXYwaNYoPP/yQpUuXMnPmTMaNG8f8+fPZv38/Dz300Cm/zuVyAfDiiy8SHR3dbZvb7b7gcYuIiGjwExER6YGysjL+/ve/s3z5clavXs3kyZP5+c9/Htj+wQcfAF2redpsNmw2W2BbXl4eAHV1dUyaNClw/yOPPEJubi433XRTkF6FiIj0VjrHT0RE5CStra3U1tZSW1vLwYMHWbNmDUuXLuXSSy/l2muvJSUlhd27d7Nz504OHjzI888/z3PPPQcQOHQzLi6O/fv3U11dTU5ODldddRUPPvggmzZt4sCBA6xYsYIXX3yR3Nxcky9VRER6CZulCw2JiIgEzJgxg8OHDwdux8XFkZmZyVVXXcXNN99MTEwM9fX1PPDAA2zbtg2Hw0F+fj433ngjy5Yt44UXXmD8+PH86U9/4ic/+Ql2u52tW7fS1tbGk08+yYYNG2hqaiI3N5c77riDyy+/3OCrFREk+MmvAAAAa0lEQVSR3kKDn4iIiIiISITToZ4iIiIiIiIRToOfiIiIiIhIhNPgJyIiIiIiEuE0+ImIiIiIiEQ4DX4iIiIiIiIRToOfiIiIiIhIhNPgJyIiIiIiEuE0+ImIiIiIiEQ4DX4iIiIiIiIR7v8DbjKYlsIa8cwAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1080x360 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Creating Y to predict\n",
    "period = 2\n",
    "ibov_df['Close'+str(period)] = ibov_df['Adj Close'].shift(-period)\n",
    "ibov_df['High'+str(period)] = ibov_df['High'].shift(-period)\n",
    "ibov_df['y'] = np.where(ibov_df['High'+str(period)] - ibov_df['Open'] > 500, 1,0)\n",
    "ibov_df['y'] = ibov_df['y'].shift(-1)\n",
    "ibov_df['Variation'] = (ibov_df['Close'+str(period)] - ibov_df['Open']).shift(-1)\n",
    "ibov_df['Perfect_y'] = ibov_df['y'] * ibov_df['Variation']\n",
    "ibov_df['Perfect_y_accum'] = ibov_df['Perfect_y'].cumsum()\n",
    "\n",
    "# Drop nan\n",
    "ibov_df.dropna(inplace=True)\n",
    "\n",
    "# Plot Best scenario \n",
    "sns.set(rc={'figure.figsize':(15,5)})\n",
    "ibov_df['Perfect_y_accum'].plot()\n",
    "plt.xlabel('Date', fontsize=15)\n",
    "plt.ylabel('Points Accumulated', fontsize=15)\n",
    "plt.title('Best Scenario', fontsize=20)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 418,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0    1951\n",
       "1.0    1285\n",
       "Name: y, dtype: int64"
      ]
     },
     "execution_count": 418,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Statistics\n",
    "ibov_df['y'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 419,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df_features shape:  (3236, 5)\n",
      "df_target shape:  (3236,)\n"
     ]
    }
   ],
   "source": [
    "# Split Features and Target\n",
    "df_features = ibov_df[['Open', 'High', 'Low','Adj Close', 'Min_Low']].copy()\n",
    "df_target   = ibov_df['y'].copy()\n",
    "\n",
    "# Create a size for the training set \n",
    "train_size = int(0.75 * df_target.shape[0])\n",
    "\n",
    "X_train = df_features[:train_size]\n",
    "y_train = df_target[:train_size]\n",
    "X_test  = df_features[train_size:]\n",
    "y_test  = df_target[train_size:]\n",
    "print('df_features shape: ', df_features.shape)\n",
    "print('df_target shape: ', df_target.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 423,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Open</th>\n",
       "      <th>High</th>\n",
       "      <th>Low</th>\n",
       "      <th>Adj Close</th>\n",
       "      <th>Min_Low</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Datetime</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2019-04-05 13:00:00-03:00</th>\n",
       "      <td>97014.18750</td>\n",
       "      <td>97235.50000</td>\n",
       "      <td>96960.06250</td>\n",
       "      <td>97028.34375</td>\n",
       "      <td>54.12500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2019-04-05 14:00:00-03:00</th>\n",
       "      <td>97028.06250</td>\n",
       "      <td>97248.25781</td>\n",
       "      <td>97026.95313</td>\n",
       "      <td>97142.07813</td>\n",
       "      <td>1.10937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2019-04-05 15:00:00-03:00</th>\n",
       "      <td>97125.68750</td>\n",
       "      <td>97283.42969</td>\n",
       "      <td>96855.34375</td>\n",
       "      <td>96885.22656</td>\n",
       "      <td>270.34375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2019-04-05 16:00:00-03:00</th>\n",
       "      <td>96929.82813</td>\n",
       "      <td>97180.89063</td>\n",
       "      <td>96926.38281</td>\n",
       "      <td>97134.64844</td>\n",
       "      <td>3.44532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2019-04-08 10:00:00-03:00</th>\n",
       "      <td>97110.04688</td>\n",
       "      <td>97438.50000</td>\n",
       "      <td>96901.14844</td>\n",
       "      <td>97117.84375</td>\n",
       "      <td>208.89844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2020-09-11 15:00:00-03:00</th>\n",
       "      <td>98181.95313</td>\n",
       "      <td>98646.47656</td>\n",
       "      <td>98035.09375</td>\n",
       "      <td>98538.77344</td>\n",
       "      <td>146.85938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2020-09-11 16:00:00-03:00</th>\n",
       "      <td>98499.00000</td>\n",
       "      <td>98705.53906</td>\n",
       "      <td>98269.95313</td>\n",
       "      <td>98365.75781</td>\n",
       "      <td>229.04687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2020-09-14 10:00:00-03:00</th>\n",
       "      <td>98366.61719</td>\n",
       "      <td>99597.60156</td>\n",
       "      <td>98366.61719</td>\n",
       "      <td>99286.86719</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2020-09-14 11:00:00-03:00</th>\n",
       "      <td>99258.63281</td>\n",
       "      <td>99493.70313</td>\n",
       "      <td>98974.70313</td>\n",
       "      <td>99320.21875</td>\n",
       "      <td>283.92968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2020-09-14 12:00:00-03:00</th>\n",
       "      <td>99333.35156</td>\n",
       "      <td>99685.99219</td>\n",
       "      <td>99254.39063</td>\n",
       "      <td>99431.04688</td>\n",
       "      <td>78.96093</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2427 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  Open         High          Low    Adj Close  \\\n",
       "Datetime                                                                        \n",
       "2019-04-05 13:00:00-03:00  97014.18750  97235.50000  96960.06250  97028.34375   \n",
       "2019-04-05 14:00:00-03:00  97028.06250  97248.25781  97026.95313  97142.07813   \n",
       "2019-04-05 15:00:00-03:00  97125.68750  97283.42969  96855.34375  96885.22656   \n",
       "2019-04-05 16:00:00-03:00  96929.82813  97180.89063  96926.38281  97134.64844   \n",
       "2019-04-08 10:00:00-03:00  97110.04688  97438.50000  96901.14844  97117.84375   \n",
       "...                                ...          ...          ...          ...   \n",
       "2020-09-11 15:00:00-03:00  98181.95313  98646.47656  98035.09375  98538.77344   \n",
       "2020-09-11 16:00:00-03:00  98499.00000  98705.53906  98269.95313  98365.75781   \n",
       "2020-09-14 10:00:00-03:00  98366.61719  99597.60156  98366.61719  99286.86719   \n",
       "2020-09-14 11:00:00-03:00  99258.63281  99493.70313  98974.70313  99320.21875   \n",
       "2020-09-14 12:00:00-03:00  99333.35156  99685.99219  99254.39063  99431.04688   \n",
       "\n",
       "                             Min_Low  \n",
       "Datetime                              \n",
       "2019-04-05 13:00:00-03:00   54.12500  \n",
       "2019-04-05 14:00:00-03:00    1.10937  \n",
       "2019-04-05 15:00:00-03:00  270.34375  \n",
       "2019-04-05 16:00:00-03:00    3.44532  \n",
       "2019-04-08 10:00:00-03:00  208.89844  \n",
       "...                              ...  \n",
       "2020-09-11 15:00:00-03:00  146.85938  \n",
       "2020-09-11 16:00:00-03:00  229.04687  \n",
       "2020-09-14 10:00:00-03:00    0.00000  \n",
       "2020-09-14 11:00:00-03:00  283.92968  \n",
       "2020-09-14 12:00:00-03:00   78.96093  \n",
       "\n",
       "[2427 rows x 5 columns]"
      ]
     },
     "execution_count": 423,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_train"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Model Training - Random Forest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 413,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 330, in fit\n",
      "    y, expanded_class_weight = self._validate_y_class_weight(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\", line 558, in _validate_y_class_weight\n",
      "    check_classification_targets(y)\n",
      "  File \"C:\\Users\\Usuario\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\", line 172, in check_classification_targets\n",
      "    raise ValueError(\"Unknown label type: %r\" % y_type)\n",
      "ValueError: Unknown label type: 'unknown'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Unknown label type: 'unknown'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-413-53e9705a6cf0>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      6\u001b[0m \u001b[0mts\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mTimeSeriesSplit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mn_splits\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m4\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      7\u001b[0m \u001b[0mclf\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mGridSearchCV\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mestimator\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mrfr\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcv\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mts\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mparam_grid\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mgrid\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 8\u001b[1;33m \u001b[0mclf\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mX_train\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0my_train\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      9\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     10\u001b[0m \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"Tuned Random Forest Regression Parameters: {}\"\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mclf\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mbest_params_\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\sklearn\\utils\\validation.py\u001b[0m in \u001b[0;36minner_f\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m     71\u001b[0m                           FutureWarning)\n\u001b[0;32m     72\u001b[0m         \u001b[0mkwargs\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m{\u001b[0m\u001b[0mk\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0marg\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0marg\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msig\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mparameters\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0margs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m}\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 73\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mf\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     74\u001b[0m     \u001b[1;32mreturn\u001b[0m \u001b[0minner_f\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     75\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_search.py\u001b[0m in \u001b[0;36mfit\u001b[1;34m(self, X, y, groups, **fit_params)\u001b[0m\n\u001b[0;32m    763\u001b[0m             \u001b[0mrefit_start_time\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    764\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0my\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 765\u001b[1;33m                 \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mbest_estimator_\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mX\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0my\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mfit_params\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    766\u001b[0m             \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    767\u001b[0m                 \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mbest_estimator_\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mX\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mfit_params\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\u001b[0m in \u001b[0;36mfit\u001b[1;34m(self, X, y, sample_weight)\u001b[0m\n\u001b[0;32m    328\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mn_outputs_\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0my\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    329\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 330\u001b[1;33m         \u001b[0my\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mexpanded_class_weight\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_validate_y_class_weight\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0my\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    331\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    332\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0my\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"dtype\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m!=\u001b[0m \u001b[0mDOUBLE\u001b[0m \u001b[1;32mor\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0my\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mflags\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcontiguous\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\sklearn\\ensemble\\_forest.py\u001b[0m in \u001b[0;36m_validate_y_class_weight\u001b[1;34m(self, y)\u001b[0m\n\u001b[0;32m    556\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    557\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_validate_y_class_weight\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0my\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 558\u001b[1;33m         \u001b[0mcheck_classification_targets\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0my\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    559\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    560\u001b[0m         \u001b[0my\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0my\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\sklearn\\utils\\multiclass.py\u001b[0m in \u001b[0;36mcheck_classification_targets\u001b[1;34m(y)\u001b[0m\n\u001b[0;32m    170\u001b[0m     if y_type not in ['binary', 'multiclass', 'multiclass-multioutput',\n\u001b[0;32m    171\u001b[0m                       'multilabel-indicator', 'multilabel-sequences']:\n\u001b[1;32m--> 172\u001b[1;33m         \u001b[1;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"Unknown label type: %r\"\u001b[0m \u001b[1;33m%\u001b[0m \u001b[0my_type\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    173\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    174\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: Unknown label type: 'unknown'"
     ]
    }
   ],
   "source": [
    "# Hyperparameters to search\n",
    "grid = {'max_depth': [*range(2,6)], 'max_features': [*range(2,6)], 'random_state': [42]}\n",
    "\n",
    "# Find the best hyperparameters\n",
    "rfr = RandomForestClassifier(n_estimators=200)\n",
    "ts = TimeSeriesSplit(n_splits=4)\n",
    "clf = GridSearchCV(estimator=rfr, cv=ts, param_grid=grid)\n",
    "clf.fit(X_train, y_train)\n",
    "\n",
    "print(\"Tuned Random Forest Regression Parameters: {}\".format(clf.best_params_)) \n",
    "print(\"Best score is {}\".format(clf.best_score_))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performance Metrics - Random Forest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Results\n",
    "randfor_pred = clf.predict(X_test)\n",
    "print(\"R2: {}\".format(clf.score(X_test, y_test)))\n",
    "print()\n",
    "print(pd.crosstab(y_test, randfor_pred))\n",
    "print()\n",
    "print(classification_report(y_test, randfor_pred))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def not_stoped(data, period):\n",
    "    df = ibov_df[['Low', 'Open']].copy()\n",
    "    shifts = []\n",
    "    period=5\n",
    "    for i in range(1,period+1):\n",
    "        df['Low'+ str(i)] = df['Low'].shift(-i)\n",
    "        shifts.append('Low'+ str(i))\n",
    "\n",
    "    df['Min'] = df.min(axis=1)\n",
    "    \n",
    "    data['Not_Stop'] = df['Open'] - df['Min'] < 500\n",
    "    \n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "avaliation_df['RandFor_y'] * avaliation_df['Variation'] < -500 & avaliation_df['Stop']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "avaliation_df = ibov_df[['Open', 'High'+str(period), 'Close'+str(period) ,'Adj Close', 'y', 'Variation']][train_size:]\n",
    "avaliation_df['Perfect_y_accum'] = (avaliation_df['y'] * avaliation_df['Variation']).cumsum()\n",
    "not_stoped(avaliation_df,5)\n",
    "\n",
    "avaliation_df['RandFor_y'] = randfor_pred\n",
    "avaliation_df['RandFor_Result'] = np.where(avaliation_df['RandFor_y'] * avaliation_df['Variation'] < -500 & avaliation_df['Stop'] , -500,avaliation_df['RandFor_y'] * avaliation_df['Variation'])\n",
    "avaliation_df['RandFor_Result_accum'] = avaliation_df['RandFor_Result'].cumsum()\n",
    "\n",
    "#plt.plot(avaliation_df['Perfect_y_accum'], color = 'darkblue', label= 'Best Cenary')\n",
    "plt.plot(avaliation_df['RandFor_Result_accum'], color = 'orange', label= 'Random Forest')\n",
    "plt.xlabel('Date', fontsize=15)\n",
    "plt.ylabel('Points Accumulated', fontsize=15)\n",
    "plt.title('With Risk Management', fontsize=20)\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SVM Model\n",
    "svm = SVC(kernel = 'rbf',gamma='scale')\n",
    "svm.fit(X_train,y_train)\n",
    "svm_pred = svm.predict(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Results\n",
    "print(\"R2: {}\".format(svm.score(X_test, y_test)))\n",
    "print()\n",
    "print(pd.crosstab(y_test, svm_pred))\n",
    "print()\n",
    "print(classification_report(y_test, svm_pred))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate metrics\n",
    "avaliation_df['SVM_y'] = svm_pred\n",
    "avaliation_df['SVM_Result'] =  np.where(avaliation_df['SVM_y'] * avaliation_df['Variation'] < -500, -500,avaliation_df['SVM_y'] * avaliation_df['Variation'])\n",
    "avaliation_df['SVM_Result_accum'] = avaliation_df['SVM_Result'].cumsum()\n",
    "\n",
    "# Plot Results\n",
    "plt.plot(avaliation_df['RandFor_Result_accum'], color = 'orange', label= 'Random Forest')\n",
    "plt.plot(avaliation_df['SVM_Result_accum'], color = 'black', label= 'SVM')\n",
    "plt.xlabel('Date', fontsize=15)\n",
    "plt.ylabel('Points Accumulated', fontsize=15)\n",
    "plt.title('With Risk Management', fontsize=20)\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Random Forest points per months\n",
    "ibov_aval = avaliation_df.groupby(avaliation_df.index.month).sum()\n",
    "plt.bar(ibov_aval.index, ibov_aval.RandFor_Result, color='darkblue')\n",
    "plt.xlabel('Months', fontsize= 15)\n",
    "plt.ylabel('Points', fontsize= 15)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Relationship between high of the day and close of the day\n",
    "maior = avaliation_df[avaliation_df['Close'+str(period)] > avaliation_df['Open']]\n",
    "plt.hist(maior['High'+str(period)] - maior['Open'],bins = 25, label='Máximas', color='darkblue')\n",
    "plt.hist(maior['Close'+str(period)] - maior['Open'],bins = 25, color='red', label='Fechamentos', alpha=0.4)\n",
    "plt.legend()\n",
    "plt.xlabel('Points', fontsize= 15)\n",
    "plt.ylabel('Occurrency', fontsize= 15)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mean Loss and Mean Profit\n",
    "mean_profit = avaliation_df[avaliation_df['RandFor_Result']>0]['RandFor_Result'].mean()\n",
    "mean_loss   = avaliation_df[avaliation_df['RandFor_Result']<0]['RandFor_Result'].mean()\n",
    "max_profit = avaliation_df[avaliation_df['RandFor_Result']>0]['RandFor_Result'].max()\n",
    "\n",
    "print('Mean Profit:', mean_profit)\n",
    "print('Mean Loss:', mean_loss)\n",
    "print('Max Profit:', max_profit)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(avaliation_df.index, avaliation_df['Adj Close'], label='IBOV 60')\n",
    "\n",
    "b = np.where(avaliation_df['RandFor_y'] == 1, True, False)\n",
    "avaliation_df['markers_on'] = avaliation_df['Adj Close'][b]\n",
    "\n",
    "plt.plot(avaliation_df.index, avaliation_df['markers_on'], '^', markersize=5, color='g', label='IBOV 60')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "total = clf.predict(df_features)\n",
    "\n",
    "avaliation_df = ibov_df[['Low', 'Open', 'High'+str(period), 'Close'+str(period) ,'Adj Close', 'y', 'Variation']]\n",
    "avaliation_df['Perfect_y_accum'] = (avaliation_df['y'] * avaliation_df['Variation']).cumsum()\n",
    "\n",
    "avaliation_df['RandFor_y'] = total\n",
    "avaliation_df['RandFor_Result'] = np.where(avaliation_df['Low'].rolling(3).min() - avaliation_df['Open'] < -500, -500,avaliation_df['RandFor_y'] * avaliation_df['Variation'])\n",
    "avaliation_df['RandFor_Result_accum'] = avaliation_df['RandFor_Result'].cumsum()\n",
    "\n",
    "#plt.plot(avaliation_df['Perfect_y_accum'], color = 'darkblue', label= 'Best Cenary')\n",
    "plt.plot(avaliation_df['RandFor_Result_accum'], color = 'orange', label= 'Random Forest')\n",
    "plt.xlabel('Date', fontsize=15)\n",
    "plt.ylabel('Points Accumulated', fontsize=15)\n",
    "plt.title('With Risk Management', fontsize=20)\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mean Loss and Mean Profit\n",
    "mean_profit = avaliation_df[avaliation_df['RandFor_Result']>0]['RandFor_Result'].mean()\n",
    "mean_loss   = avaliation_df[avaliation_df['RandFor_Result']<0]['RandFor_Result'].mean()\n",
    "max_profit = avaliation_df[avaliation_df['RandFor_Result']>0]['RandFor_Result'].max()\n",
    "\n",
    "print('Mean Profit:', mean_profit)\n",
    "print('Mean Loss:', mean_loss)\n",
    "print('Max Profit:', max_profit)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
